### DISCUSSION TITLE
how to pass tailwind classes via props in next js?

### DISCUSSION THREAD

**Discussion by saman2007** [tailwindlabs/tailwindcss#8995] [Help] [answered]
hello. i have a problem. i have a json file with these values:
```
{
  "placeholder": "search something...",
  "textColor": "text-[#222F5D]",
  "placeHolderColor": "placeholder:text-[#222F5D]",
  "bgColor": "bg-[#EBEBEB]",
  "searchIconColor": "text-[#222F5D]"
}
```
these tailwind classes(except `placeholder`) are for my input and form element. i send these values to my input and form component as props and i use them to style them. but the problem starts here that the styles doesnt apply on the elements! it will be added to the class lists of the element but their styles dont apply on the elements. i searched a lot about my problem but i couldnt solve my problem. im even using `classNames` package but it couldnt solve my problem either. and i should say that i am using `hero icons` package. what should i do when i want to get some classes from a json file and then send them to my components as props to style my elements? here are my components that i am sending these classes to them:

```
import navItems from "../data/NavItems.json";
import searchBarDatas from "../data/SearchBar.json";

//in _app.js in the default component
<Navigation items={navItems} searchBarDatas={searchBarDatas} />
```
```
//navigation component:
import NavItems from "./NavItems";
import SearchBar from "./SearchBar";

const Navigation = ({ items, searchBarDatas }) => {
  return (
    <nav className="bg-stone-50 left-0 right-0 h-[80px] fixed shadow-[0px_4px_14px_rgba(0,0,0,0.08)] flex justify-between items-center p-[10px]">
      <SearchBar {...searchBarDatas} />
      <NavItems items={items} />
    </nav>
  );
};

export default Navigation;
```
```
//searchbar component:
import SearchButton from "./SearchButton";
import classNames from "classnames";

const SearchBar = ({
  placeholder,
  textColor,
  placeholderColor,
  bgColor,
  searchIconColor,
}) => {
  const formClasses = classNames(
    "rounded-[5px] max-w-[500px] h-[50px] flex-grow flex flex-row-reverse items-center",
    bgColor
  );

  const inputClasses = classNames(
    "bg-transparent px-[7px] py-[5px] w-full outline-none h-full text-right",
    placeholderColor,
    textColor
  );

  return (
    <form className={formClasses}>
      <SearchButton color={searchIconColor} />
      <input
        type="text"
        placeholder={placeholder}
        className={inputClasses}
        style={{ direction: "rtl" }}
      />
    </form>
  );
};

export default SearchBar;
```
```
//search button component:
import SearchIcon from "@heroicons/react/outline/SearchIcon";
import classNames from "classnames";

const SearchButton = ({ color }) => {
  return (
    <button
      className={`rounded-[5px] bg-transparent outline-none p-[10px] ${color}`}
      type="submit"
    >
      <SearchIcon
        className={classNames(
          "max-w-[25px] max-h-[25px] min-w-[25px] min-h-[25px",
          color
        )}
      />
    </button>
  );
};

export default SearchButton;
```
![image](https://user-images.githubusercontent.com/91914392/181937167-5909d75f-3b51-471c-aab4-e62237aedf29.png)
look at this picture. none of the styles have been applied! but why?

so whats my problem? how can i solve it?
thanks for helping.

**Comment by wongjn** [comment, marked as answer]
When you say `searchIconColor` doesn't work, I'm assuming you mean that the icon color is not `#222F5D`? Double-check the icon element has the `text-[#222F5D]`? If so, it doesn't seem to make sense that `textColor` works and `searchIconColor` doesn't since they are both using the same `text-[#222F5D]` class name. Are you sure that `textColor` is working on the input element? It might be getting it's actual rendered text color through inheritance from a parent element rather than the `text-[#222F5D]`. This could all point to the fact that `text-[#222F5D]` class might not be being generated in your CSS. If not, you may need to add your JSON files to the `content` config member of your `tailwind.config.js`.

**Reply by saman2007** [reply]
actually the class name will be added to the class list of the icon. but, its style that is `color: #222F5D;`  wont apply! the problem is exactly here. and yes. you are right. `textColor` doesnt work either. i didnt realize that.
but `bgColor` and `placeHolderColor` will be aplied.
i will update my question and add a picture of that so you can clearly understand the differences.

**Reply by wongjn** [reply]
This sounds like the `text-[#222F5D]` class might not be being generated in your CSS. If not, you may need to add your JSON files to the `content` config member of your `tailwind.config.js`. So that Tailwind scans these JSON files.

**Reply by saman2007** [reply]
i updated my question. i made some mistakes in my question.

**Reply by saman2007** [reply]
but using this json file is for development. at the end, i want to fetch these datas from api. so i think its not ok to add these datas to the `content` of `tailwind.config.js`

**Reply by wongjn** [reply]
I would then suggest using the CSS variables (for the placeholder) and the `style` attribute instead:

```diff
 {
   "placeholder": "search something...",
-  "textColor": "text-[#222F5D]",
+  "textColor": "#222F5D",
-  "placeHolderColor": "placeholder:text-[#222F5D]",
+  "placeHolderColor": "#222F5D",
-  "bgColor": "bg-[#EBEBEB]",
+  "bgColor": "#EBEBEB",
-  "searchIconColor": "text-[#222F5D]"
+  "searchIconColor": "#222F5D"
 }
```

```diff
 //searchbar component:
 import SearchButton from "./SearchButton";
-import classNames from "classnames";
 
 const SearchBar = ({
   placeholder,
   textColor,
   placeholderColor,
   bgColor,
   searchIconColor,
 }) => {
-  const formClasses = classNames(
-    "rounded-[5px] max-w-[500px] h-[50px] flex-grow flex flex-row-reverse items-center",
-    bgColor
-  );
+  const formClasses = "rounded-[5px] max-w-[500px] h-[50px] flex-grow flex flex-row-reverse items-center";
 
-  const inputClasses = classNames(
-    "bg-transparent px-[7px] py-[5px] w-full outline-none h-full text-right",
-    placeholderColor,
-    textColor
-  );
+  const inputClasses = "bg-transparent px-[7px] py-[5px] w-full outline-none h-full text-right placeholder:text-[color:var(--color-placeholder)]";
 
   return (
-    <form className={formClasses}>
+    <form className={formClasses} style={{ backgroundColor: bgColor }}>
       <SearchButton color={searchIconColor} />
       <input
         type="text"
         placeholder={placeholder}
         className={inputClasses}
-        style={{ direction: "rtl" }}
+        style={{
+         "--color-placeholder": placeholderColor,
+         color: textColor,
+         direction: "rtl",
+        }}
       />
     </form>
   );
 };
 
 export default SearchBar;
```

```diff
 //search button component:
 import SearchIcon from "@heroicons/react/outline/SearchIcon";
-import classNames from "classnames";
 
 const SearchButton = ({ color }) => {
   return (
     <button
-      className={`rounded-[5px] bg-transparent outline-none p-[10px] ${color}`}
+      className="rounded-[5px] bg-transparent outline-none p-[10px]"
+      style={{ color }}
       type="submit"
     >
       <SearchIcon
-        className={classNames(
-          "max-w-[25px] max-h-[25px] min-w-[25px] min-h-[25px",
-          color
-        )}
+        className="max-w-[25px] max-h-[25px] min-w-[25px] min-h-[25px]",
+        style={{ color }}
       />
     </button>
   );
 };
 
 export default SearchButton;
```

We use the `style` attribute to handle the background color and text color. We need to use arbitrary value  class with a CSS variable for the placeholder color since there is no way to target the placeholder color directly using the `style` attribute. Of course, we could have done this for background and text colors but since we are not using any modifiers with them, it is not needed. If you do end up using variants with these values, then you can follow the pattern we used for the placeholder color.

**Reply by saman2007** [reply]
Thank you for you answer! But is it bad for `SEO` to use `css in js` a lot in our elements? Im saying it because im creating an e-commerce website and ecommerce websites should be fully customizable and with this solution, im going to write a lot of `css in js`.  Whats your mind? Do you have any suggestion to make my website fully customizable? Whats the best way to create a fully customizable e-commerce website? Thanks for answering.

**Reply by wongjn** [reply]
This isn't CSS-in-JS. CSS-in-JS is typically using a library to write CSS in JavaScript, such as [JSS](https://cssinjs.org/), [Styled Components](https://styled-components.com/) or [Emotion](https://emotion.sh/). I'm not very knowledgeable on SEO but I don't think having a few inline styles here and there should affect rankings.
